File: Gold IV/14002.py
'''
문제: 가장 긴 증가하는 부분 수열 4(https://www.acmicpc.net/problem/14002)
분류: 다이나믹 프로그래밍
시간 복잡도: O(n^2)
'''
from typing import List


def solution(number: List[int]) -> int:
    answer: int = 1
    answer_idx: int = 0
    dp: List[int] = [1] * len(number)
    from_idxs: List[int] = [-1] * len(number)

    for i in range(len(number)):
        for j in range(i):
            if number[i] > number[j]:
                if dp[j] + 1 > dp[i]:
                    dp[i] = dp[j] + 1
                    from_idxs[i] = j
        if answer < dp[i]:
            answer = dp[i]
            answer_idx = i
    stack = []
    cur_idx = answer_idx
    while cur_idx != -1:
        stack.append(str(number[cur_idx]))
        cur_idx = from_idxs[cur_idx]
    return answer, stack[::-1]


def main():
    n: int = int(input())
    numbers: List[int] = list(map(int, input().split()))
    answer, answer_sequence = solution(numbers)
    print(answer)
    print(' '.join(answer_sequence))


if __name__ == '__main__':
    main()

File: Gold IV/14002.cpp
/*
문제: 가장 긴 증가하는 부분 수열 4(https://www.acmicpc.net/problem/14002)
분류: 다이나믹 프로그래밍
시간 복잡도: O(n^2)
 */
#include <iostream>
#include <vector>
#include <stack>

using namespace std;

vector<int> solution(vector<int> &numbers)
{
    vector<int> answer;
    vector<int> memo(numbers.size(), 1);
    vector<int> sourceIdx(numbers.size(), -1);
    int maxValue = 1, maxIdx = 0;
    for (int i = 1; i < numbers.size(); i++)
    {
        for (int j = 0; j < i; j++)
        {
            if (numbers[j] < numbers[i])
            {
                if (memo[j] + 1 > memo[i])
                {
                    memo[i] = memo[j] + 1;
                    sourceIdx[i] = j;
                    if (maxValue < memo[i])
                    {
                        maxValue = memo[i];
                        maxIdx = i;
                    }
                }
            }
        }
    }
    stack<int> st;
    while (maxIdx != -1)
    {
        st.push(numbers[maxIdx]);
        maxIdx = sourceIdx[maxIdx];
    }
    while (!st.empty())
    {
        answer.push_back(st.top());
        st.pop();
    }
    return answer;
}
int main()
{
    int n;
    cin >> n;
    vector<int> numbers(n);
    for (int i = 0; i < n; i++)
    {
        cin >> numbers[i];
    }
    vector<int> answer = solution(numbers);
    cout << answer.size() << '\n';
    for (int each : answer)
    {
        cout << each << ' ';
    }
    return 0;
}